packages/jest-jasmine2/src/p_cancelable.js

```diff
@@ -0,0 +1,83 @@
+'use strict';
+
+// Try getting the real promise object from the context, if available. Someone
+// could have overridden it in a test.
+const Promise =
+  (global.__originalGlobals__ && global.__originalGlobals__.Promise) ||
+  global.Promise;
+
+class CancelError extends Error {
+  constructor() {
+    super('Promise was canceled');
+    this.name = 'CancelError';
+  }
+}
+
+class PCancelable {
+  static fn(fn) {
+    return function() {
+      const args = [].slice.apply(arguments);
+      return new PCancelable((onCancel, resolve, reject) => {
+        args.unshift(onCancel);
+        fn.apply(null, args).then(resolve, reject);
+      });
+    };
+  }
+
+  constructor(executor) {
+    this._pending = true;
+    this._canceled = false;
+
+    this._promise = new Promise((resolve, reject) => {
+      this._reject = reject;
+
+      return executor(
+        fn => {
+          this._cancel = fn;
+        },
+        val => {
+          this._pending = false;
+          resolve(val);
+        },
+        err => {
+          this._pending = false;
+          reject(err);
+        },
+      );
+    });
+  }
+
+  then() {
+    return this._promise.then.apply(this._promise, arguments);
+  }
+
+  catch() {
+    return this._promise.catch.apply(this._promise, arguments);
+  }
+
+  cancel() {
+    if (!this._pending || this._canceled) {
+      return;
+    }
+
+    if (typeof this._cancel === 'function') {
+      try {
+        this._cancel();
+      } catch (err) {
+        this._reject(err);
+      }
+    }
+
+    this._canceled = true;
+    this._reject(new CancelError());
+  }
+
+  get canceled() {
+    return this._canceled;
+  }
+}
+
+Object.setPrototypeOf(PCancelable.prototype, Promise.prototype);
```

Comment:
can we not just use `extends` above?

Reply:
Using `extends` forces you to call `super` on the constructor, which is something the original author did not want (this file is copy-pasted from `p-cancelable`'s NPM module, since I need it to also use the real `Promise` object and there's no way of injecting it).